# Classification
Author: Ruben Schild<br>
Student number: 650580<br>
Date: 21-5-2022

For this notebook to work you must have installed the following packages (usually via pip install *packageName*:
* numpy
* pandas
* **\<add other packages\>**

From these we will need the following libraries:

In [1]:
# pandas as pd
try:
    import pandas as pd
    print('Pandas already installed, only imported')
except:
    !pip install pandas
    import pandas as pd
    print('Pandas was not installed, installed and imported')
    
# numpy as np
try:
    import numpy as np
    print('NumPy already installed, only imported')
except:
    !pip install numpy
    import numpy as np
    print('NumPy was not installed, installed and imported')
    
# pyplot as plt
try:
    import matplotlib.pyplot as plt
    print('PyPlot already installed, only imported')
except:
    !pip install matplotlib
    import matplotlib.pyplot as plt
    print('PyPlot was not installed, installed and imported')    

# sns

try: 
    import seaborn as sns
    print('seaborn already installed, only imported')
except:
    !pip install seaborn
    print('seaborn was not installed, installed and imported')

try:
    import statsmodels.api as sm
    print('statsmodels already installed, only imported')

except:
    !pip install statsmodels
    print('statsmodels was not installed, installed and imported')

# sklearn
try:
    from sklearn.linear_model import LinearRegression
    print('sklearn already installed, only imported')
except:
    !pip install sklearn
    from sklearn.linear_model import LinearRegression
    print('sklearn was not installed, installed and imported')

from sklearn import metrics

try:
    import eli5
    print('eli5 already installed, only imported')
except:
    !pip install eli5
    import eli5
    print('eli5 was not installed, installed and imported')

from eli5.sklearn import PermutationImportance


Pandas already installed, only imported
NumPy already installed, only imported
PyPlot already installed, only imported
seaborn already installed, only imported
statsmodels already installed, only imported
sklearn already installed, only imported
eli5 already installed, only imported


# 1. The Data
We are going to use the datafile from a customers personality analysis. The collected data is the buying behaviour of customers at a grocery store.

In [2]:
# enter the code to load the data
loan_df = pd.read_csv('../data/loan.csv')


Lets see if there are missing values:

In [3]:
loan_df.isna().sum()


Loan ID                           514
Customer ID                       514
Loan Status                       514
Current Loan Amount               514
Term                              514
Credit Score                    19668
Annual Income                   19668
Years in current job             4736
Home Ownership                    514
Purpose                           514
Monthly Debt                      514
Years of Credit History           514
Months since last delinquent    53655
Number of Open Accounts           514
Number of Credit Problems         514
Current Credit Balance            514
Maximum Open Credit               516
Bankruptcies                      718
Tax Liens                         524
dtype: int64

Looks like there are some some missing values, so lets remove them:

In [4]:
loan_df.dropna(inplace=True)
loan_df.isna().sum()


Loan ID                         0
Customer ID                     0
Loan Status                     0
Current Loan Amount             0
Term                            0
Credit Score                    0
Annual Income                   0
Years in current job            0
Home Ownership                  0
Purpose                         0
Monthly Debt                    0
Years of Credit History         0
Months since last delinquent    0
Number of Open Accounts         0
Number of Credit Problems       0
Current Credit Balance          0
Maximum Open Credit             0
Bankruptcies                    0
Tax Liens                       0
dtype: int64

Here's a quick look at the data:

In [5]:
# enter the code to see the first few rows of the data
loan_df.head()


Loan ID  \
2   4eed4e6a-aa2f-4c91-8651-ce984ee8fb26   
6   273581de-85d8-4332-81a5-19b04ce68666   
8   8af915d9-9e91-44a0-b5a2-564a45c12089   
10  32c2e48f-1ba8-45e0-a530-9a6622c18d9c   
12  403d7235-0284-4bb6-919a-09402fecbf7b   

                             Customer ID Loan Status  Current Loan Amount  \
2   5efb2b2b-bf11-4dfd-a572-3761a2694725  Fully Paid           99999999.0   
6   90a75dde-34d5-419c-90dc-1e58b04b3e35  Fully Paid             217646.0   
8   af534dea-d27e-4fd6-9de8-efaa52a78ec0  Fully Paid             548746.0   
10  0de7bcdb-ebf4-4608-ba39-05f083f855b6  Fully Paid           99999999.0   
12  11581f68-de3c-49d8-80d9-22268ebb323b  Fully Paid           99999999.0   

          Term  Credit Score  Annual Income Years in current job  \
2   Short Term         741.0      2231892.0              8 years   
6   Short Term         730.0      1184194.0             < 1 year   
8   Short Term         678.0      2559110.0              2 years   
10  Short Term         728.0       714628.0              3 years   
12  Short Term         740.0       776188.0             < 1 year   

   Home Ownership             Purpose  Monthly Debt  Years of Credit History  \
2        Own Home  Debt Consolidation      29200.53                     14.9   
6   Home Mortgage  Debt Consolidation      10855.08                     19.6   
8            Rent  Debt Consolidation      18660.28                     22.6   
10           Rent  Debt Consolidation      11851.06                     16.0   
12       Own Home  Debt Consolidation      11578.22                      8.5   

    Months since last delinquent  Number of Open Accounts  \
2                           29.0                     18.0   
6                           10.0                     13.0   
8                           33.0                      4.0   
10                          76.0                     16.0   
12                          25.0                      6.0   

    Number of Credit Problems  Current Credit Balance  Maximum Open Credit  \
2                         1.0                297996.0             750090.0   
6                         1.0                122170.0             272052.0   
8                         0.0                437171.0             555038.0   
10                        0.0                203965.0             289784.0   
12                        0.0                134083.0             220220.0   

    Bankruptcies  Tax Liens  
2            0.0        0.0  
6            1.0        0.0  
8            0.0        0.0  
10           0.0        0.0  
12           0.0        0.0

In [6]:
# Code to see the types of loan statuses
loan_df['Loan Status'].unique()


array(['Fully Paid', 'Charged Off'], dtype=object)

**<span style ='background:yellow'>\<explain the variables (/fields/columns) you are going to use for your prediction. The dependent y (the one you are going to predict) and two independent x's (the ones you are going to use to predict that y). What do these variables mean?\></span>**

We are going to predict the status of a loan based on the anual income and the monthly debt the borrower has. <br>
These independent variables should be interesting on the outcome of the loan status. A high annual income would probably mean you can fully pay off the loan whereas a low anual income it would be more difficult. <br>
The same goes for monthly debt, but then the other way around; if the borrower has a higher debt per month, the chance of the loan being charged off could be higher than when there is a lower monthly debt. 

*note:* Its easiest if your independent x variables are numeric.

In [7]:
# Make a separate datafram based on the selected variables
loan_ss = loan_df[['Loan Status', 'Annual Income', 'Monthly Debt']]

In [8]:
loan_ss.isna().sum()

Loan Status      0
Annual Income    0
Monthly Debt     0
dtype: int64

The categorical dependent variable loan status has the following categories:

In [9]:
# code that returns the different categories in the y variable.
loan_ss['Loan Status'] = pd.Categorical(loan_ss['Loan Status'])
loan_ss['Loan Status'].cat.categories


C:\Users\Ruben\AppData\Local\Temp\ipykernel_1752\4021867281.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  loan_ss['Loan Status'] = pd.Categorical(loan_ss['Loan Status'])


Index(['Charged Off', 'Fully Paid'], dtype='object')

We'll also need some training and testing data, so:

In [10]:
# code to split data in training and testing
from sklearn.model_selection import train_test_split

# Set input and output
X = loan_ss[['Annual Income', 'Monthly Debt']]
y = loan_ss['Loan Status']

#Split the data
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2,random_state=109)

All set, lets try to predict this using our independent variable Annual Income

# 2. The Basic Classification Models

In the Jupyter Notebook from lecture 5 a few different Classification techniques were discussed. Lets explore how these perform on the data.

Before we look at some 'real' models, its a good idea to get a baseline in by using one or more of the dummy classifiers. Lets see how they perform:

In [11]:
# code to create, fit and measure the dummy classifiers (see chapter 5.4. in the lecture notebook)
# include both the accuracy score and the confusion matrix for each.
from sklearn.dummy import DummyClassifier
from sklearn.metrics import confusion_matrix

# Most frequent
dumMF = DummyClassifier(strategy='most_frequent')
dumMF = dumMF.fit(X_train, y_train)
y_pred = dumMF.predict(X_test)
print(metrics.accuracy_score(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))


0.7964310226492793
[[   0 1483]
 [   0 5802]]


In [12]:
# Unified
dumUni = DummyClassifier(strategy="uniform")
dumUni.fit(X_train, y_train)
y_pred = dumUni.predict(X_test)
print(metrics.accuracy_score(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))

0.5018531228551819
[[ 734  749]
 [2880 2922]]


**<span style ='background:yellow'>\<explain all the results. What do the numbers mean?></span>**

So the first dummy model performed quite well actually, with a score 80% rounded up with the testing data. The second model performed quite bad with a score of 51% rounded up <br>
The confusion matrix of the first model predicted 0 times the annual income and it was 0 times. It predicted 5802 the monthly debt and it was the monthly debt. 1483 it predicted montly debt but was annual income. And 0 times it predicted annual income and it was 0 times annual income.<br>
The confusion matrix of the second model predicted 746 times annual income and that was true, it predicted 2969 times monthly debt and that was also true. But it predicted 737 times monthly debt but it was annual income. And it predicted 2833 times the annual income but was monthly debt.

So, those are our 'baseline'. A model should be able to at least outperform these.

Lets dive in...

## 2.1. Naive Bayes

The first model discussed was the Naive Bayes model. 

**\<explain briefly in your own words how a Naive Bayes method works>** <br>
The Naive Bayes theorem assumes that there is independence among the other predictors. With this theorem you can calculate $P(A|B)$ if B is known, given A. So $P(B|A)$



Lets create and fit this model:

In [37]:
# code to create the model, and fit the data.
from sklearn.naive_bayes import GaussianNB

# Because we use numerical input, we need the Gaussian Distribution
gaussNB = GaussianNB()

# Make the model and create a prediction
gaussNB.fit(X_train, y_train)
y_predG = gaussNB.predict(X_test)


Now to measure its performance:

In [38]:
# code to show its accuracy score AND confusion matrix.
print(metrics.accuracy_score(y_test, y_predG))
print(confusion_matrix(y_test, y_predG))

0.7964310226492793
[[   0 1483]
 [   0 5802]]


**<span style ='background:yellow'>\<explain all the results. What do the numbers mean? How is this compared to the dummy classifiers?></span>**

Lets also have a look at what a prediction would be. If the **<span style ='background:yellow'>\<your first x variable><span>** has a score of **<span style ='background:yellow'>\<enter some value></span>** and the **<span style ='background:yellow'>\<your other x variable></span>** has a score of **<span style ='background:yellow'>\<enter some value></span>**, then this model will predict:

In [15]:
# code to show the prediction


That's about it for NB. A nice thing about NB is that it doesn't really require any parameters. Lets look at our next technique.

## 2.2. Support Vector Machines
The second model discussed were Support Vector Machines. There is a plural here, because we can use different kernels.

**<span style ='background:yellow'>\<explain briefly in your own words how a SVM method works></span>**

The basic kernel is the linear one, so we'll attempt that first:

In [16]:
# code to create the model, and fit the data.


Measuring its performance....

In [17]:
# code to show its accuracy score AND confusion matrix.


**<span style ='background:yellow'>\<explain all the results. What do the numbers mean? How is this compared to the dummy classifiers, and the NB?></span>**

Lets do the same for the other kernels that were discussed, i.e. rbf, polynomial, and sigmoid.

In [18]:
# code to create the models, fit the data, and show its accuracy score AND confusion matrix.
# make sure to print some text between to indicate which result belongs to which model.


**<span style ='background:yellow'>\<explain all the results. What do the numbers mean? How is this compared to the dummy classifiers?></span>**

Allright, lets move on to the third technique...

## 2.3. K-Nearest Neighbors
The third technique is the K-Nearest Neighbors (KNN). 

**<span style ='background:yellow'>\<explain briefly in your own words how a KNN method works></span>**

To use this we need to do some additional steps.

First we need to normalize our x variables:

In [19]:
# use the StandardScaler to normalize the two x variables


Second we need to determine how many neighbors (k) we want. To do this we'll visualize the results using different values for k.

In [20]:
# Code to create the graph with Error Rate vs. K-values.


**<span style ='background:yellow'>\<explain the result and your choice for k based on the graph></span>**

*Note:* +0.5 if you also use the GridSearch technique to decide on k.

Now that we know what we want k to be, we can create the model.

In [21]:
# code to create the model with the selected k


Lets find out how good it is:

In [22]:
# code to show its accuracy score AND confusion matrix.


**<span style ='background:yellow'>\<explain all the results. What do the numbers mean? How is this compared to the dummy classifiers, the NB, and the SVM kernels?></span>**

One more basic technique to go.

## 2.4. Decision Trees
The last technique that was discussed in detail, were the Decision Trees. 

**<span style ='background:yellow'>\<explain briefly in your own words how a Decision Tree method works></span>**

The following variations were discussed:

* ID3 (or entropy with sklearn)
* Gini
* Random Forest
* Extra trees

Hopefully we have the hang of this now, so lets do each of them in one go:

In [23]:
# code to create the models, fit the data, and show its accuracy score AND confusion matrix.
# make sure to print some text between to indicate which result belongs to which model.


**<span style ='background:yellow'>\<explain all the results. What do the numbers mean? How is this compared to the dummy classifiers, the NB, the SVM kernels, and the knn?></span>**

One last set of techniques to explore.

## 2.5. Other Models
In the Jupyter Notebook from the lecture, in chapter 5.4 a few more techniques were simply shown:

* Linear Discriminant Analysis
* Quadratic Discriminant Analysis
* Logistic Regression Classifier
* Multinomial Logistic Regression Classification
* Adaptive Boosting
* Gradient Boosting
* Histogram Gradient Boosting
* XGBoost
* Stacking

Out of curiousity lets see how these perform.

In [24]:
# code to create the models, fit the data, and show its accuracy score (the confusion matrix is here optional).
# make sure to print some text between to indicate which result belongs to which model.


**<span style ='background:yellow'>\<which performed best?></span>**

<font size="5"><center>-----Chapters 1 and 2 are required to be fully completed to get a 60, the next few chapters will give a +10 for each chapter.<br> 
    However the template is not as extensive as the previous chapters. <br>
    You can select any chapter below the order is not fixed (you can leave the others empty)<br>
    You don't have to use the same dataset for the chapters below. If it helps in clarification you can use another dataset, but then make sure to include it as you submit.
    ----</center></font>

# 3. Visualisation

With two input parameters we can actually determine visually where a model will classify a variable into which category. An overview of such plots is shown at https://scikit-learn.org/stable/auto_examples/classification/plot_classifier_comparison.html

We cannot copy that code since it does a comparison. What we want is a function that takes the X and Y data as input, as well as the model to be used and then shows the decision areas.

In [25]:
# the code for the function


In [26]:
# an example of using the function


# 4. Categorical Input
With classification we have a categorical output variable, but what if we also have one or more categorical input variables.

One popular technique is one-hot-encoding, but there are others.

In this chapter we'll discuss **<span style ='background:yellow'>\<your chosen technique></span>**.

**<span style ='background:yellow'>\<explain the technique in detail. What does it do and how does it work></span>**

In [27]:
# example code of using this technique


# 5. Performance
Some models get higher accuracy scores than others. In the Jupyter Notebook from the lecture the UFC data was used and the QDA had the highest accuracy score: 0.6747. The big question is, can it be done better? First areas to look for improvement are to simply increase the number of input variables, or tweak some parameters of some of the models, or a combination of both.

In this chapter we'll give it an attempt.

First we need to load the data:

In [28]:
# code to load the UFC data


**<span style ='background:yellow'>\<explain your attempt, what did you do.></span>**

In [29]:
# code that will generate an accuracy score for the outcome that is higher than 0.6747


# 6. A New Technique

Many techniques were discussed in class and the lecture Jupyter Notebook, but there are a lot more. In this chapter the \<your chosen new technique> is discussed.

**<span style ='background:yellow'>\<explain in detail this new technique. Note that other students should be able to understand it from your explanation alone!></span>**

In [30]:
# code on using this technique


**<span style ='background:yellow'>\<feel free to use more cells for this, you probably need them></span>**